На ветке lesson4

RamzanBersanov opened the pull request:
Изменения, которые будут включены в коммит: задание 4, дубль 1

RamzanBersanov committed:
На ветке lesson4
 Изменения, которые будут включены в коммит: задание 4, дубль 1

VidgarVii commented on app/models/answer.rb:
```diff
@@ -1,2 +1,3 @@
 class Answer < ApplicationRecord
+  belongs_to :question
```
с ассоциациями ты поторопился, это будет чуть дальше в курсе
в этом задании нужно все решать через SQL, с помощью  ActiveRecord queries, но без использования ассоциаций
суть урока - понять какую работу делает за нас `ActiveRecord` и для этого нужно здесь выполнить ее самому
давай сделаем следующее:
- избавься от ассоциаций (оставь связи только на уровне БД, но не на уровне моделей). Всё ровно они объявлены не корректно
- перепиши методы в моделях `User`, `Test` без использования ассоциаций

VidgarVii commented on db/seeds.rb:
```diff
@@ -1,7 +1,56 @@
+Test.create!([{
+  title: "History",
+  level: "1",
+  author_id: "1",
```
"хардкодить" `id` - не очень хорошая затея
такой "сид" невозможно накатить на существующую базу, ведь там id могут быть заняты другими записями или вовсе отсутствовать
правильней воспользоваться тем, что метод `create` возвращает созданные записи

VidgarVii commented on db/seeds.rb:
```diff
@@ -1,7 +1,56 @@
+  correct: true,
+  question_id: "3"
+}])
+p "Created #{Answer.count} Answers"
```
мало данных. Тестировать будет сложно. Тем более нет вариантов ответов.
Для каждого теста хорошо бы сделать по 3-4 вопроса. Для каждого вопроса, по 3-4 ответа

RamzanBersanov committed:
На ветке lesson4
 Изменения, которые будут включены в коммит: сиды, убраны ассоциации

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,9 @@
 class Test < ApplicationRecord
+
```
подобные отсупы не нужны

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,9 @@
+  def self.by_category(category_title)
+    category = Category.where(title: category_title).first
+    where(category_id: category.id)
+    # joins(:category).where(category: {title: category}).order(title: :desc)
```
отладочный или старый код надо удалять, а не комментировать. Сейчас это просто мусор

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,9 @@
+
+  def self.by_category(category_title)
+    category = Category.where(title: category_title).first
+    where(category_id: category.id)
```
Так конечно тоже будет работать.
Но на много эффективней будет использовать join таблиц.
1 Тут будет 2 запроса в бд. С join 1 запрос
2 Рельса часто может оптимизировать код в 1 запрос. Так называемый ленивое выполнение запроса (но мне кажется это не тот случай) И тут получится под запрос. Под запросы всегда работают медленней чем join таблиц

VidgarVii replied:
ТЗ: "Создайте метод класса в модели Test, который возвращает **отсортированный по убыванию массив названий всех Тестов** у которых Категория называется определённым образом (название категории передается в метод в качестве аргумента)."

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,2 +1,6 @@
 class User < ApplicationRecord
+
+  def passed_tests(test_level)
+    Test.joins('JOIN tests_users ON tests.id = tests_users.test_id').where('tests_users.user_id = ? AND tests.level = ?', self.id, test_level)
```
можно сделать метод более читаемым вызывая методы с новой строки
```ruby
Test
  .joins('JOIN tests_users ON tests.id = tests_users.test_id')
  .where('tests_users.user_id = ? AND tests.level = ?', self.id, test_level)
```
в where лучше будет смотреться хеш синтаксис
```ruby
where(level: test_level, tests_users: { user_id: id })
```

VidgarVii commented on db/migrate/20221101194616_create_questions.rb:
```diff
@@ -2,7 +2,7 @@ class CreateQuestions < ActiveRecord::Migration[7.0]
     create_table :questions do |t|
       t.text :body, null: false
-      t.references :test, null: false, foreign_key: true
+      t.references :test, foreign_key: true
```
интересно, какая причина убирать констрейн null: false
По логике нам не нужны вопросы без тестов и ответы без вопросов

VidgarVii commented on db/migrate/20221104235022_create_tests_users.rb:
```diff
@@ -0,0 +1,11 @@
+    create_table :tests_users do |t|
+      t.references :test, foreign_key: true, null: false
+      t.references :user, foreign_key: true, null: false
+      t.boolean :complete, default: false
```
для boolean полей всегда рекомендуется устанавливать null: false
чтобы возможных значений было только 2 true/false

VidgarVii commented on db/seeds.rb:
```diff
@@ -1,7 +1,85 @@
+logic_answer3_q2 = Answer.create!(body: "Кочан капусты", correct: false, question_id: logical_question2.id)
+logic_answer1_q3 = Answer.create!(body: "Разговор", correct: true, question_id: logical_question3.id)
+logic_answer2_q3 = Answer.create!(body: "Шнурки", correct: false, question_id: logical_question3.id)
+logic_answer3_q3 = Answer.create!(body: "Наушники", correct: false, question_id: logical_question3.id)
```
Если переменная не используется, то лучше её не создавать. Затрудняет работу сборщику мусора

RamzanBersanov committed:
На ветке lesson4
 Изменения, которые будут включены в коммит:
	изменено:      app/models/test.rb
	изменено:      app/models/user.rb
	изменено:      db/migrate/20221101194616_create_questions.rb
	изменено:      db/migrate/20221104235022_create_tests_users.rb
	изменено:      db/seeds.rb

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,8 @@
 class Test < ApplicationRecord
+  def self.by_category(category_title)
+    joins('JOIN categories ON categories.id = tests.category_id')
+    .where('categories.title = ?', category_title)
```
при переносе вызовов методов используют отступы
```ruby
joins('JOIN categories ON categories.id = tests.category_id')
  .where('categories.title = ?', category_title)
  .order(title: :desc)
```
так понятней, что это единный запрос

VidgarVii replied:
про хеш синтаксис я уже писал 
https://github.com/RamzanBersanov/testguru/pull/4#discussion_r1014602668

VidgarVii replied:
и еще раз читаем очень внимательно задание. Что должен возвращать метода?
https://github.com/RamzanBersanov/testguru/pull/4#discussion_r1014602408

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,2 +1,6 @@
 class User < ApplicationRecord
+  def passed_tests(test_level)
+    Test.joins('JOIN tests_users ON tests.id = tests_users.test_id')
+        .where('tests_users.user_id = ? AND tests.level = ?', self.id, test_level)
```
так же не исправлены аргументы для where

VidgarVii commented on db/migrate/20221101194543_create_tests.rb:
```diff
@@ -3,8 +3,8 @@ def change
-      t.references :author, null: false, foreign_key: { to_table: :users }
-      t.references :category, null: false, foreign_key: true
+      t.references :author, foreign_key: { to_table: :users }
+      t.references :category, foreign_key: true
```
не исправлено 
https://github.com/RamzanBersanov/testguru/pull/4#discussion_r1014602856

VidgarVii commented on db/migrate/20221101194654_create_answers.rb:
```diff
@@ -3,7 +3,7 @@ def change
       t.text :body, null: false
       t.boolean :correct, default: true, null: false
-      t.references :question, null: false, foreign_key: true
+      t.references :question, foreign_key: true
```
не исправлено
https://github.com/RamzanBersanov/testguru/pull/4#discussion_r1014602856

RamzanBersanov committed:
На ветке lesson4
 Изменения, которые будут включены в коммит:
	изменено:      app/models/test.rb
	изменено:      app/models/user.rb
	изменено:      db/migrate/20221101194543_create_tests.rb
	изменено:      db/migrate/20221101194654_create_answers.rb

VidgarVii commented on app/models/test.rb:
```diff
@@ -1,2 +1,7 @@
+    joins('JOIN categories ON categories.id = tests.category_id')
+      .where(categories: {title: category_title})
+      .order(title: :desc)
+  end
```
и опять же этот метод не возвращает, то что требуется в задании
 **массив названий всех Тестов**
Тебе понятно это выражение?

VidgarVii commented on db/seeds.rb:
```diff
@@ -1,7 +1,86 @@
+Answer.create!(body: "Наушники", correct: false, question_id: logical_question3.id)
+p "Created #{Answer.count} Answers"
+
+TestsUser.create(test_id: '1', user_id: '1')
```
https://github.com/RamzanBersanov/testguru/pull/4#discussion_r1012525566

RamzanBersanov committed:
На ветке lesson4
 Изменения, которые будут включены в коммит:
	изменено:      app/models/test.rb

RamzanBersanov committed:
На ветке lesson4
 Изменения, которые будут включены в коммит:
	изменено:      db/schema.rb
	изменено:      db/seeds.rb
